### PR TITLE
Lock snippets when scheduled for publishing if `LockableMixin` is applied

### LINKED ISSUE #9733: Make `LockableMixin` applicable to snippets and add locking for scheduled publishing
The workflows feature makes use of the `LockableMixin` and Wagtail's `WorkflowLock` object to lock the page when it's awaiting review. The locking mechanism in workflows itself doesn't actually make use of the database fields in `LockableMixin`, as it only uses the mixin as an interface by returning the `WorkflowLock` object in `get_lock()`.

The `LockableMixin` is currently not applicable to non-page models because the `related_name` for the `user` `ForeignKey` is hardcoded to `locked_pages`. Even though we'll only need the mixin for the `get_lock()` method, it's worth making the mixin applicable to snippets as well so that we can just reuse the mixin instead of extracting the `get_lock()` method into a separate class. This also allows us to implement the locking feature for snippets.

### PR DESCRIPTION
**Pull request by laymonage**
Fixes #9733. Needs #9709.

Will add docs when the manual locking feature has been implemented.

_Please check the following:_

-   [x] Do the tests still pass?[^1]
-   [x] Does the code comply with the style guide?
    -   [x] Run `make lint` from the Wagtail root.
-   [x] For Python changes: Have you added tests to cover the new/fixed behaviour?

**Please describe additional details for testing this change**.

- Apply `LockableMixin` to a snippet model that has `DraftStateMixin` applied and `PublishingPanel` exists in the panel configuration.
- Schedule an instance of the snippet to be published.
- The snippet should be locked and it should be unavailable for editing.
- Bypassing the client-side validation (e.g. by calling `form.submit()` from the JS console) would result in an error message saying that the snippet can't be saved as it is locked.

[^1]: [Development Testing](https://docs.wagtail.org/en/latest/contributing/developing.html#testing)
[^2]: [Browser and device support](https://docs.wagtail.org/en/latest/contributing/developing.html#browser-and-device-support)
[^3]: [Accessibility Target](https://docs.wagtail.org/en/latest/contributing/developing.html#accessibility-targets)

### CODE REVIEW

**Review by gasman — CHANGES_REQUESTED** [review]

File: wagtail/locks.py
Comment (laymonage) [reply wagtail/locks.py:146]: Ah yes, I wasn't quite finished with this one. I avoided checking `isinstance(self.object, Page)` because I didn't like the idea of this module importing `wagtail.models` (which also imports this module). I'll put the import in the method then, as I guess it's unavoidable in this case...

Somewhat related, I was hoping to raise some questions in the core team meeting regarding `get_admin_display_title` and also how we should handle/avoid issues in translatable UI strings for snippets (e.g. using model's `verbose_name` while there are languages with gendered nouns). And about the possibility of just using the same strings for pages and snippets... I'll put that on the agenda for tomorrow.

Comment (laymonage) [reply wagtail/locks.py]: Note: please do not re-review the PR yet, there are some more tests I'd like to add. I'll request another review when it's ready.

**Review by laymonage — COMMENTED** [review]
I think this should be ready for another review @gasman. Thanks!

**Review by gasman — APPROVED** [review]

**Comment by gasman** [comment]
Thanks @laymonage! Will hold off on adding a release note until the documentation is in place.

**Pull request merged**
